File: android/lib/src/main/java/me/anasmusa/learncast/lib/component/ConfirmationBottomSheet.kt
package me.anasmusa.learncast.lib.component

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import me.anasmusa.learncast.Strings
import me.anasmusa.learncast.string

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ConfirmationBottomSheet(
    title: String? = null,
    message: String? = null,
    positiveButtonTitle: String? = null,
    negativeButtonTitle: String? = null,
    onConfirm: () -> Unit,
    onDismiss: () -> Unit,
) {
    val sheetState =
        rememberModalBottomSheetState(
            skipPartiallyExpanded = true,
        )

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
    ) {
        Column(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .padding(
                        start = 20.dp,
                        end = 20.dp,
                        bottom = 16.dp,
                    ),
            horizontalAlignment = Alignment.Start,
        ) {
            title?.let {
                Text(
                    text = title,
                    style = MaterialTheme.typography.headlineSmall,
                    fontWeight = FontWeight.SemiBold,
                    color = MaterialTheme.colorScheme.onSurface,
                )

                Spacer(modifier = Modifier.height(12.dp))
            }

            message?.let {
                Text(
                    text = message,
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurface,
                )
            }

            Spacer(modifier = Modifier.height(24.dp))

            SheetMenuWhiteButton(
                modifier = Modifier.fillMaxWidth(),
                title = positiveButtonTitle ?: Strings.YES.string(),
                icon = null,
                horizontalArrangement = Arrangement.Center,
                onClick = {
                    onConfirm()
                },
            )

            Spacer(modifier = Modifier.height(12.dp))

            SheetMenuButton(
                modifier = Modifier.fillMaxWidth(),
                title = negativeButtonTitle ?: Strings.CANCEL.string(),
                icon = null,
                horizontalArrangement = Arrangement.Center,
                onClick = onDismiss,
            )

            Spacer(modifier = Modifier.height(8.dp))
        }
    }
}

File: ios/lib/component/ConfirmationBottomSheet.swift
//
//  ConfirmationBottomSheet.swift
//  ios
//
//  Created by Anas Erkinjonov on 13/02/26.
//

internal import Shared
import SwiftUI

struct ConfirmationBottomSheet: View {
    @Environment(\.env) var env: AppEnvironment

    let title: String?
    let message: String?
    let positiveButtonTitle: String?
    let negativeButtonTitle: String?
    let onConfirm: () -> Void
    let onDismiss: () -> Void

    init(
        title: String? = nil,
        message: String? = nil,
        positiveButtonTitle: String? = nil,
        negativeButtonTitle: String? = nil,
        onConfirm: @escaping () -> Void,
        onDismiss: @escaping () -> Void
    ) {
        self.title = title
        self.message = message
        self.positiveButtonTitle = positiveButtonTitle
        self.negativeButtonTitle = negativeButtonTitle
        self.onConfirm = onConfirm
        self.onDismiss = onDismiss
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Spacer()
                .frame(height: 12)

            // Title
            if let title = title {
                Text(title)
                    .font(Typography.headlineSmall)
                    .fontWeight(.semibold)
                    .foregroundColor(.primary)

                Spacer()
                    .frame(height: 12)
            }

            // Message
            if let message = message {
                Text(message)
                    .font(Typography.bodyMedium)
                    .foregroundColor(.primary)
            }

            Spacer()
                .frame(height: 24)

            // Positive Button
            SheetMenuWhiteButton(
                icon: nil,
                title: positiveButtonTitle ?? Strings.shared.YES_TEXT.string(),
                horizontalAlignment: .center,
                onClick: onConfirm
            )

            Spacer()
                .frame(height: 12)

            // Negative Button
            SheetMenuButton(
                icon: nil,
                title: negativeButtonTitle ?? Strings.shared.CANCEL.string(),
                horizontalAlignment: .center,
                onClick: onDismiss
            )

            Spacer()
                .frame(height: 8)
        }
        .padding(.horizontal, 20)
        .presentationDetents([.height(calculateHeight())])
        .presentationBackground(env.backgroundGradient())
        .presentationDragIndicator(.visible)
    }

    private func calculateHeight() -> CGFloat {
        var height: CGFloat = 0

        // Title height
        if title != nil {
            height += 32 + 12  // Approximate title height + spacing
        }

        // Message height
        if message != nil {
            height += 64  // Approximate message height
        }

        // Spacing + buttons + bottom padding
        height += 24 + 56 + 12 + 56 + 8 + 16  // spacing + button1 + gap + button2 + padding + safe area

        return height
    }
}

#Preview {
    Text("Preview")
        .sheet(isPresented: .constant(true)) {
            ConfirmationBottomSheet(
                title: "Sign Out?",
                message: "Are you sure you want to sign out?",
                positiveButtonTitle: "Sign Out",
                negativeButtonTitle: "Cancel",
                onConfirm: {},
                onDismiss: {}
            )
        }
}
